session.go

```diff
@@ -312,16 +314,27 @@ func (s *Session) Ping() (dur time.Duration, err error) {
 		s.pingLock.Unlock()
 	}()
 
-	// Send the ping request
-	hdr := encode(typePing, flagSYN, 0, activePing.id)
-	if err := s.sendMsg(hdr, nil, nil); err != nil {
-		return 0, err
+	// Send the ping request, waiting at most one connection write timeout
+	// to flush it.
+	timer := time.NewTimer(s.config.ConnectionWriteTimeout)
+	defer timer.Stop()
+	select {
+	case s.pingCh <- activePing.id:
```

Comment:
Note behavior change: if the write of the ping fails, it would previously return as an error from this method. In this case it will not, as the error is not retrieved back from the other go routine.

Reply:
Sends were always async. `sendMsg` may return an error due to a shutdown and/or a timeout; those are handled internally here. If writing to the connection itself fails, the entire connection will be killed.